README.md
# pkgr

[![Build Status](https://travis-ci.org/crohr/pkgr.svg?branch=master)](https://travis-ci.org/crohr/pkgr)

## Goal

Make debian or rpm packages out of any app, including init script, crons,
logrotate, etc. Excellent way to distribute apps or command line tools without
complicated installation instructions.

Hosted service available at [Packager.io][packager-io]. Free for OpenSource apps.

[packager-io]: https://packager.io/

## Officially supported languages

* Ruby
* NodeJS
* Python
* Go

In beta:

* PHP, using the following buildpack: `https://github.com/pkgr/heroku-buildpack-php#buildcurl`

You can also point to other buildpacks
([doc](https://packager.io/documentation/customizing-the-build/#buildpack)).
They may just work.

## Supported distributions (64bits only)

* Ubuntu 22.04 ("jammy")
* Ubuntu 20.04 ("focal")
* Ubuntu 18.04 ("bionic")
* Ubuntu 16.04 ("xenial")
* Ubuntu 14.04 ("trusty")
* Ubuntu 12.04 ("precise")
* Debian 11 ("bullseye")
* Debian 10 ("buster")
* Debian 9 ("stretch")
* Debian 8 ("jessie")
* Debian 7 ("wheezy")
* RHEL/CentOS 8
* RHEL/CentOS 7
* RHEL/CentOS 6
* Suse Linux Enterprise Server 12
* Suse Linux Enterprise Server 11
* Amazon Linux AMI 2015
* Amazon Linux AMI 2014

## Examples

See [Packager.io][packager-io] for examples of apps packaged with `pkgr` (Gitlab, OpenProject, Discourse, etc.).

## Installation

Docker images are available for the most recent distributions. If you don't
have docker or are interested in older distributions, you can also install
`pkgr` as a ruby gem:

    gem install pkgr

## Usage

To package your app, execute `pkgr` against your app's repository:

    # if using the ruby gem, you need to run it under the distribution you want to package for:
    pkgr package path/to/app/repo

    # if using a docker image, you can run it from any host OS:
    docker run --rm -it -v $(pwd):/app -v /tmp/cache:/cache pkgr/ubuntu:20.04
    docker run --rm -it -v $(pwd):/app -v /tmp/cache:/cache pkgr/el:8
    docker run --rm -it -v $(pwd):/app -v /tmp/cache:/cache pkgr/debian:10

The resulting `.deb` or `.rpm` package will be in your current working directory. 

Full command line options are given below:

    $ pkgr help package
    Usage:
      pkgr package TARBALL|DIRECTORY

    Options:
      [--buildpack=BUILDPACK]                        # Custom buildpack to use
      [--buildpack-list=BUILDPACK_LIST]              # Specify a file containing a list of buildpacks to use (--buildpack takes precedence if given)
      [--changelog=CHANGELOG]                        # Changelog
      [--maintainer=MAINTAINER]                      # Maintainer
      [--vendor=VENDOR]                              # Package vendor
      [--architecture=ARCHITECTURE]                  # Target architecture for the package
                                                     # Default: x86_64
      [--runner=RUNNER]                              # Force a specific runner (e.g. upstart-1.5, sysv-lsb-1.3)
      [--logrotate-frequency=FREQUENCY]              # Set logrotate frequency
                                                     # Default: daily
                                                     # Possible values: daily, weekly, monthly, yearly
      [--logrotate-backlog=BACKLOG]                  # Set logrotate backlog
                                                     # Default: 14
      [--homepage=HOMEPAGE]                          # Project homepage (e.g. "https://pkgr.example.org")
      [--home=HOME]                                  # Project home (e.g. "/usr/share/PACKAGE_HOME")
      [--description=DESCRIPTION]                    # Project description
      [--category=CATEGORY]                          # Category this package belongs to
                                                     # Default: none
      [--version=VERSION]                            # Package version (if git directory given, it will use the latest git tag available)
      [--iteration=ITERATION]                        # Package iteration (you should keep the default here)
                                                     # Default: 20141015024539
      [--license=LICENSE]                            # The license of your package (see <https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/#license-short-name>)
      [--user=USER]                                  # User to run the app under (defaults to your app name)
      [--group=GROUP]                                # Group to run the app under (defaults to your app name)
      [--compile-cache-dir=COMPILE_CACHE_DIR]        # Where to store the files cached between packaging runs. Path will be resolved from the temporary code repository folder, so use absolute paths if needed.
      [--before-precompile=BEFORE_PRECOMPILE]        # Provide a script to run just before the buildpack compilation, on the build machine. Path will be resolved from the temporary code repository folder, so use absolute paths if needed.
      [--after-precompile=AFTER_PRECOMPILE]          # Provide a script to run just after the buildpack compilation, on the build machine. Path will be resolved from the temporary code repository folder, so use absolute paths if needed.
      [--before-install=BEFORE_INSTALL]              # Provide a script to run just before a package gets installated or updated, on the target machine.
      [--after-install=AFTER_INSTALL]                # Provide a script to run just after a package gets installated or updated, on the target machine.
      [--before-remove=BEFORE_REMOVE]                # Provide a script to run just before a package gets uninstallated, on the target machine.
      [--after-remove=AFTER_REMOVE]                  # Provide a script to run just after a package gets uninstallated, on the target machine.
      [--dependencies=one two three]                 # Specific system dependencies that you want to install with the package
      [--build-dependencies=one two three]           # Specific system dependencies that must be present before building
      [--host=HOST]                                  # Remote host to build on (default: local machine)
      [--auto], [--no-auto]                          # Automatically attempt to install missing dependencies
      [--clean], [--no-clean]                        # Automatically clean up temporary dirs
                                                     # Default: true
      [--edge], [--no-edge]                          # Always use the latest version of the buildpack if already installed
                                                     # Default: true
      [--env=one two three]                          # Specify environment variables for buildpack (--env "CURL_TIMEOUT=2" "BUNDLE_WITHOUT=development test")
      [--force-os=FORCE_OS]                          # Force a specific distribution to build for (e.g. --force-os "ubuntu-12.04"). This may result in a broken package.
      [--store-cache], [--no-store-cache]            # Output a tarball of the cache in the current directory (name: cache.tar.gz)
      [--verify], [--no-verify]                      # Verifies output package
                                                     # Default: true
      [--data-dir=DATA_DIR]                          # Custom path to data directory. Can be used for overriding default templates, hooks(pre-, post- scripts), configs (buildpacks, distro dependencies), environments, etc.
                                                     # Default: ./pkgr/data
      [--directories=DIRECTORIES]                    # Recursively mark a directory as being owned by the package
      [--verbose], [--no-verbose]                    # Run verbosely
      [--debug], [--no-debug]                        # Run very verbosely
      [--name=NAME]                                  # Application name (if directory given, it will default to the directory name)
      [--buildpacks-cache-dir=BUILDPACKS_CACHE_DIR]  # Directory where to store the buildpacks
                                                     # Default: /home/vagrant/.pkgr/buildpacks

## Why?

Tools such as [Capistrano](http://capistranorb.com/) are great for deploying
applications, but the deployment recipe can quickly become a mess, and scaling
the deployment to more than a few servers can prove to be difficult. Plus, if
you're already using automation tools such as
[Puppet](http://www.puppetlabs.com/) to configure your servers, you have to
run two different processes to configure your infrastructure.

`pkgr` builds on top of the Heroku tools to provide you with an easy way to package you app as a debian package. The great advantage is that once you've built it and you tested that it
works once, you can deploy on any number of servers at any time and you're
sure that it will just work. Then, you can upgrade/downgrade or uninstall the whole application in one command.

Finally, it's a great way to share your open source software with your users and clients. Much easier than asking them to install all the dependencies manually! I'm in the process of making sure `pkgr` is feature complete by trying to package as many successful open-source projects as I can. Don't hesitate to test it on your app and report your findings!

## What this does

* Uses Heroku buildpacks to embed all the dependencies related to your application runtime within the debian package. For a Rails app for instance, this means that `pkgr` will embed the specific ruby runtime you asked for, along with all the gems specified in your Gemfile. However, all other dependencies you may need must be specified as additional system dependencies (see Usage). This avoids the 'packaging-the-world' approach used by other tools such as omnibus (with the pros and cons that come with it), but it still allows you to use the latest and greatest libraries for your language of choice. See this [blog post][background-pkgr] for more background.

[background-pkgr]: https://blog.packager.io/posts/1-why-i-made-packager-io-digressions-on-software-packaging

* Gives you a nice executable, which closely replicates the Heroku toolbelt utility. For instance, assuming you're packaging an app called `my-app`, you can do the following:

        my-app config:set VAR=value
        my-app config:get VAR
        my-app run [procfile process] # e.g. my-app run rake db:migrate; my-app run console; etc.
        my-app run [arbitrary process] # e.g. my-app run ruby -v; my-app run bundle install; etc.
        my-app scale web=1 worker=1
        my-app logs [--tail]
        ...

* Your app will reside in `/opt/app-name`.

* You'll also get upstart, systemd, or sysvinit initialization scripts (depending on your distribution) that you can use directly:

        service my-app start/stop/restart/status

  Note: init scripts are generated only after you've done a `scale` command for the process type, e.g. `my-app scale web=1`.

* Logs will be stored in `/var/log/app-name/`, with a proper logrotate config automatically added. For systemd-based distributions, you will find the logs in the systemd journal.

* Config files can be added in `/etc/app-name/`

## Requirements

* You must have a Procfile.

* Your application should be Heroku compatible, meaning you should be able to set your main app's configuration via environment variables.

## Troubleshooting

If you're on older versions of Debian, you may need to append `/var/lib/gems/1.9.1/bin` to your PATH to "see" the `pkgr` command:

    export PATH="$PATH:/var/lib/gems/1.9.1/bin"

If you get the following error `ERROR:  While executing gem ... (ArgumentError) invalid byte sequence in US-ASCII` while trying to install `pkgr`, try setting a proper locale, and then retry:

    sudo locale-gen en_US.UTF-8
    export LANG=en_US.UTF-8
    export LC_ALL=en_US.UTF-8
    sudo gem install pkgr

Looking for the init script? It is created the first time you run this command

    sudo my-app scale web=1 worker=1

Issue getting nokogiri to compile? Try the following based on this [comment](https://github.com/crohr/pkgr/issues/60#issuecomment-96631181):

    bundle config --local build.nokogiri "--use-system-libraries --with-xml2-include=/usr/include/libxml2"

## Authors

* Cyril Rohr <cyril.rohr@gmail.com> - <http://crohr.me>, <https://packager.io>

## Copyright

See LICENSE (MIT)


CHANGELOG.md
## master

- Support SLES-15
- Pass APP_PKG_VERSION and APP_PKG_ITERATION during compilation step
- Update python buildpack to v221-1

## 1.9.0

- add support for Ubuntu 22.04
- create crons dir if missing
- update nodejs buildpack to v197
- upgrade ruby buildpack
- add --no-gpg-checks for SLES11 and SLES12 when setting up **build** dependencies. This is suboptimal, but helps with outdated package repos.

## 1.8.0

- Publish docker images to simplify build
- Upgrade ruby buildpack to v227-1
- Improve log file display

## 1.7.3

- Add support for Ubuntu 20.04
- Update ruby buildpack to v212-1

## 1.7.2

- Add support for RHEL 8.
- Update ruby buildpack to v206-1. Supports Rails6.
- Add support for Debian 10 "buster"
- Add `config:unset` to CLI.
- Update ruby buildpack to v199-1. Supports bundler2.

## 1.7.1

- Work around bundler requiring a writable home dir for executing bundle install

## 1.7.0

- Ubuntu 18.04
- Ruby buildpack v183-1

## 1.6.0

- Handle both opt/ logs and systemd logs
- Support for debian 9
- Support for configuring app environment through `cat env-file | my-app configure`
- Running `my-app configure` will try to run a configure script if one found in `packaging/scripts/configure`
- Introduction of `my-app restart [process]` command, to abstract differences among init systems
- `my-app logs` now forwards to journalctl, on systemd-enabled distributions
- Lots of fixes and improvements as to how PORTs are defined
  - Interactive runs properly handle PORT given on command line, e.g. `PORT=xxx my-app run web`
  - A PORT is now assigned for each Procfile process, offset by 100 from base PORT for each process type
  - Each worker of a specific process type is assigned a PORT equal to BASE_PORT+PORT_OFFSET+WORKER_INDEX-1
  - Existing PORT configuration is not touched for backwards compatibility (you only get the new configuration when running the `scale` command).
- Permit specifying multiple buildpacks to execute
- Upgrade default ruby and nodejs buildpacks to v164 and v104
- Fix runtime postgres dependency for sles12
- Add home parameter in CLI
- Do not attempt to create /usr/bin if CLI disabled
- Ensure existence of the logs directory
- Travis updates and build status image

## 1.5.1 - 20160908

- Update Ruby buildpack URL to get the latest ruby versions asutomatically
- Add support for Python!

## 1.5.0 - 20160908

- Fix upstart job for CentOS
- Reinstall master and process master init scripts whenever scale is called
- Run chroot with the user's configured groups
- Add SLES11 support
- Fix postinstall shebang
- Can now disable default dependencies with `default_dependencies: false`
- Allow to override `data_dir` folder
- Generate init files with correct PORT
- Allow hard-coded defaults to be overridden by pkgr.yml
- Pass ENV_DIR to compile script
- Replace only instances of `/app/` and not `/app` in buildpacks
- Create /app dir, which is required by some buildpacks to compile (e.g. GO)
- Pass TARGET as an env variable when compiling, so that it can be used to fetch binaries from buildcurl.com
- Add possibility to set category and directories fpm option
- Support for Ubuntu 16.04
- Add command-line flags for disabling default dependencies and CLI

## 1.4.4 - 20150512

- Add CentOS / RHEL 7 support with systemd
- Add Debian 8 "jessie" support with systemd
- Upstart: make sure to start once filesystem is up
- Add Amazon Linux 2015 support
- Allow addons to be installed from a local relative path
- Add SLES12 support
- Extract SVN version number, if available
- Changed interpreter in maintainer scripts to /bin/bash in order to avoid forbidden-postrm-interpreter errors on Debian 7

## 1.4.3 - 20150304

- Allow to run some commands as the APP_USER instead of root
- Some profile.d scripts output stuff (e.g. newest nodejs buildpack), so redirect to /dev/null

## 1.4.2 - 20150227

- Use Shellwords to escape command line args
- Allow scaledowns to finish even if stopping the service does not work
- Retry the packaging command at most 3 times if the package verification failed
- Add CLI options for `--before-remove` and `--after-remove` scripts
- Make sure `.git` directories are not included in the resulting package

## 1.4.1 - 20150217

- Allow to set config variables with equal signs in their values
- Add --vendor option
- Add net-tools to dependencies when using installer
- Add which to dependencies when using installer
- Add support for SLES12 (rpm)
- Get remote compiling working
- All processes from Procfile are now exported as potential services
- Add option to verify the generated packages
- CLI support: allow to overwrite default CLI to point to custom executable

## 1.4.0 - 20141015

- Add support for Fedora 20.
- Add support for packaging Go apps.
- Add support for installing cron files automatically.
- Allow to output a compressed version of the compile cache.
- Update ruby buildpack to use universal branch of https://github.com/pkgr/heroku-buildpack-ruby.
- Preliminary support for installer wizards. New configure and reconfigure CLI commands.

## 1.3.2 - 20140527

- Added more relaxed curl timeouts for Ubuntu Trusty ruby buildpack.

## 1.3.1 - 20140506

- Add buildpacks for Ubuntu 14.04.
- Add support for pre/post install files.

## 1.3.0 - 20140502

- CLI: Support for Ubuntu Trusty 14.04
- Put the CLI in /usr/bin.
- Properly set HOME environment variable to /home/:user
- Add more relaxed timeouts on ruby buildpacks.
- Add PROCESS_MANAGER env variable if custom runner is forced.
- Fix issues with .pkgr.yml custom runner not being picked up
- Allow to force a custom runner (upstart, sysvinit).
- Handles termination of processes that fork upon starting.
- Fix sysvinit script.
- Fix permissions on /etc/appname.
- Fix dependencies installation.
- Make user:group the owner of /etc/appname/\*
- Add tests for CLI
- CentOS experimental support
- Added --after-precompile and --license options.
- Move slow test to integration.
- Outputs buildpack cloning step.


lib/pkgr/config.rb
require 'ostruct'
require 'yaml'
require 'pathname'
require 'base64'

module Pkgr
  class Config < OpenStruct
    DISTRO_COMPATIBILITY_MAPPING = {
      "ubuntu-lucid" => "ubuntu-10.04",
      "ubuntu-precise" => "ubuntu-12.04",
      "debian-squeeze" => "debian-6",
      "debian-wheezy" => "debian-7",
      "debian-jessie" => "debian-8"
    }

    class << self
      def load_file(path, distribution)
        config = YAML.load_file(path) || {}
        Pkgr.debug "Configuration from file: #{config.inspect} - Distribution: #{distribution.inspect}."

        targets = config.delete("targets") || {}

        # backward compatibility
        DISTRO_COMPATIBILITY_MAPPING.each do |from, to|
          if targets.has_key?(from)
            targets[to] = targets.delete(from)
          end
        end

        distro_config = targets[distribution.to_s]
        if distro_config.is_a?(Hash)
          distro_config.each do |k,v|
            config[k] = v
          end
        end

        self.new(config)
      end
    end

    def sesame
      binding
    end

    def merge(other)
      new_config = self.class.new
      self.each{|k,v|
        new_value = case v
        when Array
          v | (other.delete(k) || [])
        else
          v
        end
        new_config.send("#{k}=".to_sym, new_value)
      }
      other.each{|k,v| new_config.send("#{k}=".to_sym, v)}
      new_config
    end

    def each
      @table.each do |k,v|
        next if v.nil?
        yield k, v
      end
    end

    def delete(key)
      @table.delete(key)
    end

    def safe_name
      name.gsub("-", "_")
    end

    def cli?
      if disable_cli.nil?
        @table.has_key?(:cli) ? @table[:cli] : true
      else
        !disable_cli
      end
    end

    def skip_default_dependencies?
      if disable_default_dependencies.nil?
        @table[:default_dependencies] === false
      else
        disable_default_dependencies == true
      end
    end

    def home
      @table[:home] || "/opt/#{name}"
    end

    def logrotate_frequency
      @table[:logrotate_frequency] || "daily"
    end

    def logrotate_backlog
      @table[:logrotate_backlog] || 14
    end

    def user
      @table[:user] || name
    end

    def group
      @table[:group] || user
    end

    def tmpdir
      @table[:tmpdir]
    end

    def architecture
      @table[:architecture] || "x86_64"
    end

    def homepage
      @table[:homepage] || "http://example.com/no-uri-given"
    end

    def description
      @table[:description] || "No description given"
    end

    def maintainer
      @table[:maintainer] || "<someone@pkgr>"
    end

    def vendor
      @table[:vendor] || "pkgr <https://github.com/crohr/pkgr>"
    end

    def env
      @table[:env].is_a?(Pkgr::Env) ? @table[:env] : Pkgr::Env.new(@table[:env])
    end

    def buildpacks
      @table[:buildpack].is_a?(String) ? @table[:buildpack].split(",") : @table[:buildpack]
    end

    def valid?
      @errors = []
      @errors.push("name can't be blank") if name.nil? || name.empty?
      @errors.push("version can't be blank") if version.nil? || version.empty?
      @errors.push("version must start with a digit") if !version.empty? && version !~ /^\d/
      @errors.push("iteration can't be blank") if iteration.nil? || iteration.empty?
      @errors.push("user can't be blank") if user.nil? || user.empty?
      @errors.push("group can't be blank") if group.nil? || group.empty?
      @errors.empty?
    end

    def errors
      @errors ||= []
    end

    def installer
      return nil if @table[:installer].nil? || @table[:installer] == false
      @table[:installer]
    end

    def wizards
      @wizards ||= (@table[:wizards] || []).map do |wizard_string|
        wizard_string.split(/\s*\|\s*/).map do |wizard|
          if wizard.start_with?('.')
            wizard = Pathname.new(source_dir).join(wizard).realpath
          end

          Addon.new(wizard)
        end
      end
    end

    def crons
      @table[:crons] || []
    end

    def before_hook
      if before_precompile.nil? || before_precompile.empty?
        before_steps = self.before || []

        if before_steps.empty?
          nil
        else
          tmpfile = Tempfile.new(["before_hook", ".sh"])
          before_steps.each{|step| tmpfile.puts step}
          tmpfile.close
          tmpfile.path
        end
      else
        before_precompile
      end
    end

    def after_hook
      if after_precompile.nil? || after_precompile.empty?
        after_steps = self.after || []

        if after_steps.empty?
          nil
        else
          tmpfile = Tempfile.new(["after_hook", ".sh"])
          after_steps.each{|step| tmpfile.puts step}
          tmpfile.close
          tmpfile.path
        end
      else
        after_precompile
      end
    end

    def before_install
      return nil if @table[:before_install].nil?
      Pathname.new(source_dir).join(@table[:before_install]).realpath.to_s
    end

    def after_install
      return nil if @table[:after_install].nil?
      Pathname.new(source_dir).join(@table[:after_install]).realpath.to_s
    end

    def before_remove
      return nil if @table[:before_remove].nil?
      Pathname.new(source_dir).join(@table[:before_remove]).realpath.to_s
    end

    def after_remove
      return nil if @table[:after_remove].nil?
      Pathname.new(source_dir).join(@table[:after_remove]).realpath.to_s
    end

    # TODO: DRY this with cli.rb
    def to_args
      args = [
        "--name \"#{name}\"",
        "--version \"#{version}\"",
        "--user \"#{user}\"",
        "--group \"#{group}\"",
        "--iteration \"#{iteration}\"",
        "--homepage \"#{homepage}\"",
        "--home \"#{home}\"",
        "--architecture \"#{architecture}\"",
        "--description \"#{description}\"",
        "--maintainer \"#{maintainer}\"",
        "--vendor \"#{vendor}\""
      ]
      args.push "--dependencies #{dependencies.map{|d| "\"#{d}\""}.join}" unless dependencies.nil? || dependencies.empty?
      args.push "--build-dependencies #{build_dependencies.map{|d| "\"#{d}\""}.join}" unless build_dependencies.nil? || build_dependencies.empty?
      args.push "--compile-cache-dir \"#{compile_cache_dir}\"" unless compile_cache_dir.nil? || compile_cache_dir.empty?
      args.push "--before-precompile \"#{before_precompile}\"" unless before_precompile.nil? || before_precompile.empty?
      args.push "--after-precompile \"#{after_precompile}\"" unless after_precompile.nil? || after_precompile.empty?
      args.push "--before-install \"#{before_install}\"" unless before_install.nil? || before_install.empty?
      args.push "--after-install \"#{after_install}\"" unless after_install.nil? || after_install.empty?
      args.push "--before-remove \"#{before_remove}\"" unless before_remove.nil? || before_remove.empty?
      args.push "--after-remove \"#{after_remove}\"" unless after_remove.nil? || after_remove.empty?

      args.push "--license \"#{license}\"" unless license.nil? || license.empty?
      args.push "--buildpack \"#{buildpack}\"" unless buildpack.nil? || buildpack.empty?
      args.push "--buildpack_list \"#{buildpack_list}\"" unless buildpack_list.nil? || buildpack_list.empty?
      args.push "--force-os \"#{force_os}\"" unless force_os.nil? || force_os.empty?
      args.push "--runner \"#{runner}\"" unless runner.nil? || runner.empty?
      args.push "--logrotate-frequency \"#{logrotate_frequency}\"" unless logrotate_frequency.nil? || logrotate_frequency.empty?
      args.push "--logrotate-backlog \"#{logrotate_backlog}\"" unless logrotate_backlog.nil?
      args.push "--env #{env.variables.map{|v| "\"#{v}\""}.join(" ")}" if env.present?
      args.push "--auto" if auto
      args.push "--verbose" if verbose
      args.push "--store-cache" if store_cache
      args.push "--debug" if debug
      args.push "--verify" if verify
      args.push "--no-clean" if !clean
      args.push "--no-edge" if !edge
      args.push "--tmpdir" if !tmpdir
      args
    end
  end
end
